pypika/terms.py

```diff
@@ -678,9 +689,7 @@ def for_(self, table):
         :return:
             A copy of the field with it's table value replaced.
         """
-        self.args = [param.for_(table) if hasattr(param, 'for_')
```

Comment:
Wouldn't this condition here result in just using the instance of ValueWrapper since it doesn't implement `for_` ? Looks to me like this condition intended to implement this behavior already. What about it was broken?

Reply:
criterions were broken because for_ is not implemented. without this fix, my test would fail.

this condition is only on Function. I simply cleaned it up because it is not needed anymore. I felt like it was cleaner to implement on the base class than to have to implement this check on every Term object.

Reply:
I mean the assumption is that if `for_` is not implemented then you don't need to replace the table for any contained terms

Reply:
I think the issue is in the implementation of `BasicCriterion.for_`

```
    @builder
    def for_(self, table):
        self.left = self.left.for_(table)
        self.right = self.right.for_(table)
```

Looks like here it does not check if for is implemented as it does in `Term`, although i see your point, that you would have to check everywhere if it is implemented or not.

Reply:
Going to approve and merge your changes

Reply:
What if you have something like Criterion(self.table.a > 1) 

you want to do Criterion.for_(new_table). this will crash because the value wrapper doesn't implement for_

Reply:
sorry if i wasn't clear. thanks!